### ISSUE TITLE
Improve accessibility of drop–downs

### ISSUE THREAD

**Issue by oleq** [ckeditor/ckeditor5#1037]
1.  The screen reader should properly announce that the button has a popup, (#12258)
    1.  ~The screen reader should explain the way to open it, (this will also be handled by #12258)~
2.  ~The screen reader focus should move properly to the drop–down and follow the navigation across its content~ (#11838)
3.  ~Exiting the drop–down should also be explained by the text reader.~ (not needed as [discussed later](https://github.com/ckeditor/ckeditor5/issues/1037#issuecomment-1147652455))

**Comment by oleq** [comment]
1. Our dropdowns have no `aria-expanded` but should have (see an example from w3c https://www.w3.org/TR/wai-aria-practices-1.1/examples/listbox/listbox-collapsible.html)
   * The only places that have that attribute are CollapsibleView (list properties: start index, reversed section) and SplitButtonView (a button for dropdowns with an “extra arrow”).

**Comment by wimleers** [comment]
AFAICT this overlaps with #11568, #11041, #10368, #3184, #2000, #3215, #3185, #3176 and #11838.

What seems _different_ about this ticket is that it seems to stress the _screen reader_ aspect.

So I propose to retitle/rescope this to `Improve screenreader support of dropdowns`, and to do this _after_ it's been made keyboard-accessible. Doing that first is necessary anyway, and may solve some of the problems outlined here 😊

**Comment by Comandeer** [comment]
> The screen reader should properly announce that the button has a popup,

It can be done by adding the [`[aria-haspopup]`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-haspopup) and [`[aria-expanded]`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded) attributes to buttons triggering dropdowns. Depending on what the dropdown is, the `[aria-haspopup]` attribute can take different values instead of generic `true` (e.g. `menu` if the dropdown is some kind of a menu). We also need to ensure that the dropdown itself has an appropriate role (e.g. `[role=menu]` in case of menu dropdown). ARIA Practices contains a decent [menu button sample](https://www.w3.org/WAI/ARIA/apg/example-index/menu-button/menu-button-actions.html).

> The screen reader should explain the way to open it,

Using the combination of `[aria-haspopup]` and `[aria-expanded]` attributes should be enough for that.

> The screen reader focus should move properly to the drop–down and follow the navigation across its content,

As @wimleers pointed out, keyboard accessibility will be covered in other issues and should resolve the issue with screen reader focus.

> Exiting the drop–down should also be explained by the text reader.

IMO it does not need to be if exiting the dropdown can be done by a standard means (e.g. pressing <kbd>Esc</kbd>) and the focus is moved back to the button that opened the dropdown. If really needed, this behavior can be described in the accessibility help dialog (#1014).

**Comment by wimleers** [comment]
FYI: tracking this at https://www.drupal.org/project/drupal/issues/3283804 😊 

> IMO it does not need to be if exiting the dropdown can be done by a standard means

Agreed: having a screen reader speak _too much_ is not a great experience either!

**Comment by mlewand** [comment]
Closing this umbrella issue as the only one open part is ckeditor/ckeditor5#12258.
